### ISSUE TITLE
Some images convert too dark when using lossy compression

### ISSUE THREAD

**Issue by a-528** [libjxl/libjxl#2070]
**Describe the bug**
When converting JPEG images to JXL with lossy compression, some pictures will convert too dark. I've included the files and examples. I've only experienced this with some older exports from Lightroom 6.14, although I haven't tested on that many images from different sources. 

When converting back to JPEG with _djxl_, pictures again gain their original gamma/brightness.

**To Reproduce**
I don't know, I've only experienced this with older exports I still have saved as JPEGs. I've tried opening the JXL files with XnViewMP ver. 1.4.0 and ImageGlass 8.7.11.6.

**Expected behavior**
The pictures should keep the same gamma/brightness.

**Screenshots**
![Comparison](https://user-images.githubusercontent.com/122736352/212548917-0756e259-2c2b-4b36-9522-e8763c5b603d.png)

**Environment**
 - OS: Windows 10
 - cjxl/djxl version string:  cjxl v0.7.0 f95da13 [AVX2,SSE4,SSSE3,Unknown] _**AND**_ cjxl v0.8.0 fadac12 [AVX2,SSE4,SSSE3,Unknown]

Attachments:
[Files (JPG + JXL).zip](https://github.com/libjxl/libjxl/files/10420513/Files.JPG.%2B.JXL.zip)
[Comparison Picture.zip](https://github.com/libjxl/libjxl/files/10420514/Comparison.Picture.zip)

**Comment by mo271** [comment]
Thanks for the report, @a-528! I'm not entirely sure what is going on.
Here is the jxlinfo for those two images:

```
[I] ~/libjxl ((a56ba92b…))> ./build/tools/jxlinfo output\ -d\ 00.jxl 
JPEG XL file format container (ISO/IEC 18181-2)
JPEG XL image, 4685x3123, (possibly) lossless, 8-bit RGB
Color space: 560-byte ICC profile, CMM type: "ADBE", color space: "RGB ", rendering intent: 0
JPEG bitstream reconstruction data available
Brotli-compressed Exif metadata: 15791 compressed bytes
Brotli-compressed xml  metadata: 283 compressed bytes
[I] ~/libjxl ((a56ba92b…))> ./build/tools/jxlinfo output\ -d\ 10.jxl 
JPEG XL file format container (ISO/IEC 18181-2)
JPEG XL image, 4685x3123, lossy, 8-bit RGB
Color space: 560-byte ICC profile, CMM type: "ADBE", color space: "RGB ", rendering intent: 0
```

When looking at the two jxl images in chrome they have the correct brightness and look like the original.

Considering that also djxl does the right thing with these images, I would suspect that this is a problem with the image viewers (thanks for checking with two different image viewers!). If this is the case, consider opening bugs there.

Can you produce such images with a current version of cjxl?

**Comment by eugenesvk** [comment]
I can repro this with the .8 latest version, but it's indeed likely some viewer bug

For example, on Windows, using the latest 9.0.3 beta of ImageGlass viewer, there is almost no difference in the image, the lossy JPEG-XL version is just slightly darker (this are pics converted with `0.8` cjxl):
- Top left cut is JPEG-XL `d0.5 j0`
- Everything else is JPEG-XL `d0 j0`

<img width="180" alt="←↑jxl d0 5 →↓jxl d0" src="https://user-images.githubusercontent.com/12956286/214824116-27970042-c09f-4ac3-a5a7-86077279461a.png">

However, in a Chrome-based browser there is not even this difference, the pics do look identical

But the __thumbnails__ are bugged (these are based on the pics attached in the post)
  - bottom left JPEG-XL `d0`
  - bottom center JPEG `reference`
  - bottom right JPEG-XL `d10` __much darker__

Thumbnails are usually generated with a "worse" pipeline, so it's not uncommon that they lack proper color profile adjustments

<img width="465" alt="←jxld-0 ↑•2jpg-ref ↓→3jxl-d10" src="https://user-images.githubusercontent.com/12956286/214822540-c220df45-092b-40ed-bdfc-e97dd2014b6b.png">

(these kind of gamma mistakes happens when ICC/color management in the viewer is not properly set up, but I'm still curious why this happens — I was hoping that with the absolute color space within JPEG-XL these kind of issues would be left in the past)

**Comment by tianwen2976** [comment]
I feel that libjxl has a BUG. The original image tiger.jpeg is losslessly compressed to tiger-lossless.jxl, and then decompressed to tiger2.jpeg. Also, lossy compression to tiger-q85.jxl, and then decompression to tiger3.jpeg.

It is found that tiger2.jpeg and tiger3.jpeg are different, and tiger3.jpeg is darker. The difference is between lossy and lossless compression

cjxl v0.8.0 adfc39bc [SSE4,SSSE3,Unknown]
djxl v0.8.0 adfc39bc [SSE4,SSSE3,Unknown]

commands:

cjxl -q 100  tiger.jpeg   tiger-lossless.jxl
cjxl -q 85 --lossless_jpeg=0  tiger.jpeg   tiger-q85.jxl
djxl tiger-lossless.jxl   tiger2.jpeg
djxl tiger-q85.jxl   tiger3.jpeg

tiger.jpeg tiger2.jpeg tiger3.jpeg are shown below:

![thumbnail_33feb24d28dd3643aaf4f34e791cdc0b](https://user-images.githubusercontent.com/8772352/216031725-6288bf82-c5aa-41bf-80d5-85f52ee54465.jpeg)

**Comment by eugenesvk** [comment]
But you're most likely checking with a viewer without color management.
- The tiger3.jpeg should have a different color profile `RGB_D65_SRG_Rel_Lin` vs 
- tiger.jpeg `Adobe RGB (1998)`,

The tiger3's profile becomes dark if you the viewer doesn't take this into account and falls back to the default (which happens to be the same as adobe's one)

I just did the same test, if I disable 'Use ICC profile', the image is darker, if I enable it, they look the same

ICC Off
<img width="272" alt="iccOFF" src="https://user-images.githubusercontent.com/12956286/216035339-e142f6b6-4a31-4473-8276-722e2153bd7e.png">

ICC On
<img width="272" alt="iccON" src="https://user-images.githubusercontent.com/12956286/216035351-2de1e8e9-91b0-421f-a118-dbf3aa0cfb26.png">

**Comment by tianwen2976** [comment]
I agree with your point of view, but there are indeed some image viewers that have problems, and there are relatively many, such as Pale Moon browser. further, can the compatibility be made better? thanks

**Comment by tianwen2976** [comment]
I agree with your point of view, @eugenesvk . but there are indeed some image viewers that have problems, and there are relatively many, such as Pale Moon browser. further, can the compatibility be made better? thanks

**Comment by tianwen2976** [comment]
Oh, it may be that the color profile RGB_D65_SRG_Rel_Lin is not very common, I use RGB_D65_SRG_Rel_SRG, and currently all image viewers do not have the problem of dark.

djxl --color_space=RGB_D65_SRG_Rel_SRG tiger-q85.jxl tiger3.jpeg

**Comment by eugenesvk** [comment]
Interesting, not sure why the default color profile is linear instead of the more common srg that would allow good fallback for the many non-color-managed app, let's wait for the devs to respond

**Comment by foundObjects** [comment]
> **Describe the bug**
> When converting JPEG images to JXL with lossy compression, some pictures will convert too dark. I've included the files and examples. I've only experienced this with some older exports from Lightroom 6.14, although I haven't tested on that many images from different sources. 
> 
> When converting back to JPEG with _djxl_, pictures again gain their original gamma/brightness.
> 
> **To Reproduce**
> I don't know, I've only experienced this with older exports I still have saved as JPEGs. I've tried opening the JXL files with XnViewMP ver. 1.4.0 and ImageGlass 8.7.11.6.
> 
> **Expected behavior**
> The pictures should keep the same gamma/brightness.
> 

I believe I've run into this behavior several months ago in imageglass and Tachiyomi while testing the output from a bulk comics conversion script I wrote.

Check the color space of your affected images, they're probably using ARGB. Some viewer tools don't seem to be handling color correction when a wider-than-sRGB-gamut profile is embedded. I didn't have time to babysit bug reports to the various viewers that exhibited the problem so I just crushed everything to sRGB in passing before jxl conversion.

Anyway, check your affected images for embedded wide-gamut color profiles and then double check the same images in an image tool with known good color profile handling. Krita should render colors properly, maybe GIMP does as well. I know ImageGlass was problematic the last time I checked.

**Comment by sobrus** [comment]
Hi! I'm encountering similar problems and this seems to be a bug in libjxl handling of AdobeRGB files. 
To make long story short, it uses invalid color space definition and transfer function - and even when viewed with correct color management, all wide gamut information is lost on decode.
I've put my findings in similar thread here:
https://github.com/libjxl/libjxl/issues/2550

**Comment by gitoss** [comment]
> It does not affect Display P3 wide gamut images, as they are handled properly.

Using Win64 latest binary "JPEG XL encoder v0.9.0 4b8aad2 [SSE4,SSSE3,SSE2]" and "JPEG XL decoder v0.9.0 4b8aad2 [SSE4,SSSE3,SSE2]", I cannot get Display P3 to work either (trying to work around the known Adobe RGB bug) when converting back to png/jpg.

I'm using XnViewMP and its built-in for viewing and it's built-in color management.

1. Export PNG from Lightroom: looks fine (unmanaged picture colors somewhat off as to be expected)
2. Convereted to JXL with cjxl lossy: looks fine (unamanaged picture very dark)
3. Converted back to png with djxl: color management doesn't recognize the profile, managed looks like unmanaged (just like unmanaged jxl).

... debug output:

1. exiftool original png
> [ICC_Profile:ICC-header] ProfileCMMType: Apple Computer Inc.
> [ICC_Profile:ICC-header] ProfileVersion: 4.0.0
> [ICC_Profile:ICC-header] ProfileClass: Display Device Profile
> [ICC_Profile:ICC-header] ColorSpaceData: RGB
> [ICC_Profile:ICC-header] ProfileConnectionSpace: XYZ
> [ICC_Profile:ICC-header] ProfileDateTime: 2015:10:14 13:08:57
> [ICC_Profile:ICC-header] ProfileFileSignature: acsp
> [ICC_Profile:ICC-header] PrimaryPlatform: Apple Computer Inc.
> [ICC_Profile:ICC-header] CMMFlags: Not Embedded, Independent
> [ICC_Profile:ICC-header] DeviceManufacturer: Apple Computer Inc.
> [ICC_Profile:ICC-header] DeviceModel:
> [ICC_Profile:ICC-header] DeviceAttributes: Reflective, Glossy, Positive, Color
> [ICC_Profile:ICC-header] RenderingIntent: Perceptual
> [ICC_Profile:ICC-header] ConnectionSpaceIlluminant: 0.9642 1 0.82491
> [ICC_Profile:ICC-header] ProfileCreator: Apple Computer Inc.
> [ICC_Profile:ICC-header] ProfileID: e5bb0e9867bd46cd4bbe446ebd1b7598
> [ICC_Profile] ProfileDescription: Display P3
> [ICC_Profile] ProfileCopyright: Copyright Apple Inc., 2015
> [ICC_Profile] MediaWhitePoint: 0.95045 1 1.08905
> [ICC_Profile] RedMatrixColumn: 0.51512 0.2412 -0.00105
> [ICC_Profile] GreenMatrixColumn: 0.29198 0.69225 0.04189
> [ICC_Profile] BlueMatrixColumn: 0.1571 0.06657 0.78407
> [ICC_Profile] RedTRC: (Binary data 32 bytes, use -b option to extract)
> [ICC_Profile] ChromaticAdaptation: 1.04788 0.02292 -0.0502 0.02959 0.99048 -0.01706 -0.00923 0.01508 0.75168
> [ICC_Profile]   BlueTRC: (Binary data 32 bytes, use -b option to extract)
> [ICC_Profile]   GreenTRC: (Binary data 32 bytes, use -b option to extract)`

2. jxlinfo intermediate jxl
> box: type: "JXL " size: 12
> JPEG XL file format container (ISO/IEC 18181-2)
> box: type: "ftyp" size: 20
> box: type: "jxlp" size: 242
> JPEG XL image, 3200x2400, lossy, 16-bit RGB
> num_color_channels: 3
> num_extra_channels: 0
> have_preview: 0
> have_animation: 0
> Intrinsic dimensions: 3200x2400
> Orientation: 1 (Normal)
> Color space: 548-byte ICC profile, CMM type: "appl", color space: "RGB ", rendering intent: 0
> box: type: "brob" size: 1899
> Brotli-compressed xml  metadata: 1899 compressed bytes
> box: type: "jxlp" size: 2870103

3. exiftool png after djxl
> [PNG:PNG-cICP] ColorPrimaries: BT.709
> [PNG:PNG-cICP] TransferCharacteristics: Linear
> [PNG:PNG-cICP] MatrixCoefficients: Identity matrix
> [PNG:PNG-cICP] VideoFullRangeFlag: 1

I haven't been able to recover from this by setting p3 as the default profile in XnViewMP or setting P3.icc to the png with ImageMagick...

Refering to https://github.com/libjxl/libjxl/issues/2725 and https://github.com/libjxl/libjxl/issues/2550

**Comment by jonnyawsom3** [comment]
While Linear decoding caused the output to be dark in some viewers, the root cause was an ICC issue that was previously fixed. Cannot reproduce, so removing the duplicate tag and re-closing.
